kernel: modules: define fb-io-fops module

namiltd opened the pull request:
Define define fb-io-fops module - Fbdev helpers for framebuffers in I/O memory
By creating a hidden module it is possible to load fb_io_fops.ko
according to the kernel version.

Fixes: 5048c9d ("kernel: modules: adjust the object files of kmod-fb ")
```
Package kmod-drm-radeon is missing dependencies for the following libraries:
fb_io_fops.ko
make[2]: *** [modules/video.mk:620: /__w/openwrt/openwrt/openwrt/bin/targets/x86/geode/packages/kmod-drm-radeon-6.12.28-r1.apk] Error 1
time: package/kernel/linux/compile#45.35#40.01#75.34
```

namiltd commented:
Successfully tested on x86/geode kernel 6.12.28

DragonBluep commented on package/kernel/linux/modules/video.mk:
```diff
@@ -626,7 +626,8 @@ define KernelPackage/drm-radeon
   KCONFIG:=CONFIG_DRM_RADEON
-  FILES:=$(LINUX_DIR)/drivers/gpu/drm/radeon/radeon.ko
+  FILES:=$(LINUX_DIR)/drivers/gpu/drm/radeon/radeon.ko \
+	$(LINUX_DIR)/drivers/video/fbdev/core/fb_io_fops.ko@ge6.8
```
It still a dirty fix. **CONFIG_DRM_RADEON** does not directly select **CONFIG_FB_IOMEM_FOPS**(fb_io_fops.ko).

namiltd replied:
Now as a separate module.

namiltd commented:
There is still a problem with kernel 6.12/geode

namiltd commented:
I tried to generate a separate module:  https://github.com/openwrt/openwrt/commit/d2951b72fea7efe3fef7baac0ee4b30cce09b219 and https://github.com/openwrt/openwrt/pull/18771/commits/ccae1e7262a203834149200802414c5813e42b98 but without success. There was still a problem with geode/6.12.

hauke commented:
x86/geode does not activate `CONFIG_FB_CORE`, see:
```
$ git grep CONFIG_FB_CORE target/linux/x86/
target/linux/x86/64/config-6.12:CONFIG_FB_CORE=y
target/linux/x86/64/config-6.6:CONFIG_FB_CORE=y
target/linux/x86/generic/config-6.12:CONFIG_FB_CORE=y
target/linux/x86/generic/config-6.6:CONFIG_FB_CORE=y
target/linux/x86/legacy/config-6.12:CONFIG_FB_CORE=y
target/linux/x86/legacy/config-6.6:CONFIG_FB_CORE=y
```

Please check if this option is set in the generated kernel configuration. The `CONFIG_FB_IOMEM_FOPS` option depends on `CONFIG_FB_CORE`, if it is not set it will also not set `CONFIG_FB_IOMEM_FOPS`. 

I think it would be better to put `fb_io_fops.ko` into a separate packet.

hauke commented on target/linux/bcm27xx/bcm2709/config-6.6:
```diff
@@ -187,7 +187,6 @@ CONFIG_FB_CFB_FILLRECT=y
 CONFIG_FB_CORE=y
 CONFIG_FB_DEVICE=y
-CONFIG_FB_IOMEM_FOPS=y
 CONFIG_FB_IOMEM_HELPERS=y
```
`CONFIG_FB_IOMEM_HELPERS` selects `CONFIG_FB_IOMEM_FOPS`, see https://github.com/torvalds/linux/blob/master/drivers/video/fbdev/core/Kconfig#L171
If the target config selects `CONFIG_FB_IOMEM_HELPERS=y` it should also select `CONFIG_FB_IOMEM_FOPS=y`.

namiltd replied:
I understand that I should leave CONFIG_FB_IOMEM_FOPS=y for bcm27xx?

namiltd replied:
> Done

DragonBluep commented on target/linux/loongarch64/config-6.6:
```diff
@@ -242,7 +242,6 @@ CONFIG_FB_CORE=y
 CONFIG_FB_DEVICE=y
 CONFIG_FB_EFI=y
-CONFIG_FB_IOMEM_FOPS=y
 CONFIG_FB_IOMEM_HELPERS=y
```
I would suggest not touching these symbols. They should not be manually edited.

namiltd replied:
Fixed

DragonBluep commented on target/linux/archs38/config-6.6:
```diff
@@ -93,7 +93,6 @@ CONFIG_EXT4_FS=y
 CONFIG_FAT_FS=y
 CONFIG_FB=y
 CONFIG_FB_CORE=y
-CONFIG_FB_IOMEM_FOPS=y
```
>> I would suggest not touching these symbols. They should not be manually edited.

> Fixed. I left it everywhere where CONFIG_FB_IOMEM_HELPERS is.

I mean all config files. In kernel 6.6, `CONFIG_FB_IOMEM_FOPS` is selected by `CONFIG_FB_CORE`. These changes will not make any difference.

ref: https://elixir.bootlin.com/linux/v6.6.90/source/drivers/video/fbdev/core/Kconfig#L7

namiltd replied:
Done

namiltd committed:
kernel: modules: define fb-io-fops module

Define define fb-io-fops module - Fbdev helpers for framebuffers in I/O memory
By creating a hidden module it is possible to load fb_io_fops.ko
according to the kernel version.

Fixes: 5048c9d ("kernel: modules: adjust the object files of kmod-fb ")
Package kmod-drm-radeon is missing dependencies for the following libraries:
fb_io_fops.ko
make[2]: *** [modules/video.mk:620: /__w/openwrt/openwrt/openwrt/bin/targets/x86/geode/packages/kmod-drm-radeon-6.12.28-r1.apk] Error 1
time: package/kernel/linux/compile#45.35#40.01#75.34

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18771
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Thanks! Rebased on top of main and merged!
